File: android/src/main/java/expo/modules/healthkit/ExpoHealthKitModule.kt
package expo.modules.healthkit

import android.app.Activity
import android.content.Intent
import androidx.activity.result.contract.ActivityResultContracts.RequestMultiplePermissions
import androidx.health.connect.client.PermissionController
import expo.modules.interfaces.permissions.PermissionsStatus
import expo.modules.kotlin.functions.Coroutine
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import kotlinx.coroutines.CancellableContinuation
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.suspendCancellableCoroutine
import kotlinx.coroutines.withContext
import kotlin.coroutines.resume
import kotlin.coroutines.resumeWithException

class ExpoHealthKitModule : Module() {
  private lateinit var service: HealthConnectService
  private var permissionContinuation: CancellableContinuation<Set<String>>? = null
  private val permissionContract = PermissionController.createRequestPermissionResultContract()

  override fun definition() = ModuleDefinition {
    Name("ExpoHealthKit")

    Events("onUpdate")

    OnCreate {
      val context = appContext.reactContext ?: appContext.throwingActivity.applicationContext
      service = HealthConnectService(context) { permissions ->
        requestHealthPermissions(permissions)
      }
    }

    OnActivityResult { _, payload ->
      if (payload.requestCode != REQUEST_PERMISSIONS) {
        return@OnActivityResult
      }
      val granted = permissionContract.parseResult(payload.resultCode, payload.data)
      val continuation = permissionContinuation
      permissionContinuation = null
      continuation?.resume(granted)
    }

    Function("isHealthDataAvailable") {
      HealthConnectService.isAvailable(appContext.reactContext)
    }

    AsyncFunction("requestAuthorization") Coroutine { options: Map<String, Any?> ->
      service.requestAuthorization(options)
    }

    AsyncFunction("getAuthorizationStatus") Coroutine { identifier: String ->
      service.authorizationStatus(identifier)
    }

    AsyncFunction("getRequestStatusForAuthorization") Coroutine { options: Map<String, Any?> ->
      service.requestStatusForAuthorization(options)
    }

    AsyncFunction("queryQuantitySamples") Coroutine { options: Map<String, Any?> ->
      service.queryQuantitySamples(options)
    }

    AsyncFunction("queryCategorySamples") Coroutine { options: Map<String, Any?> ->
      service.queryCategorySamples(options)
    }

    AsyncFunction("queryWorkouts") Coroutine { options: Map<String, Any?> ->
      service.queryWorkouts(options)
    }

    AsyncFunction("queryElectrocardiograms") { _: Map<String, Any?> ->
      service.unsupported("Electrocardiograms")
    }

    AsyncFunction("queryActivitySummaries") { _: Map<String, Any?> ->
      service.unsupported("Activity rings")
    }

    AsyncFunction("queryClinicalRecords") { _: Map<String, Any?> ->
      service.unsupported("Clinical records")
    }

    AsyncFunction("queryAudiograms") { _: Map<String, Any?> ->
      service.unsupported("Audiograms")
    }

    AsyncFunction("queryWorkoutRoute") { _: Map<String, Any?> ->
      service.unsupported("Workout GPS routes")
    }

    AsyncFunction("queryCorrelations") Coroutine { options: Map<String, Any?> ->
      service.queryCorrelations(options)
    }

    AsyncFunction("queryHeartbeatSeries") { _: Map<String, Any?> ->
      service.unsupported("Heartbeat series")
    }

    AsyncFunction("queryStatistics") Coroutine { options: Map<String, Any?> ->
      service.queryStatistics(options)
    }

    AsyncFunction("queryStatisticsCollection") Coroutine { options: Map<String, Any?> ->
      service.queryStatisticsCollection(options)
    }

    AsyncFunction("queryAnchored") Coroutine { options: Map<String, Any?> ->
      service.queryAnchored(options)
    }

    AsyncFunction("saveQuantitySample") Coroutine { sample: Map<String, Any?> ->
      service.saveQuantitySample(sample)
    }

    AsyncFunction("saveCategorySample") Coroutine { sample: Map<String, Any?> ->
      service.saveCategorySample(sample)
    }

    AsyncFunction("saveWorkout") Coroutine { workout: Map<String, Any?> ->
      service.saveWorkout(workout)
    }

    AsyncFunction("saveCorrelation") Coroutine { correlation: Map<String, Any?> ->
      service.saveCorrelation(correlation)
    }

    AsyncFunction("deleteObjects") Coroutine { options: Map<String, Any?> ->
      service.deleteObjects(options)
    }

    AsyncFunction("getBiologicalSex") {
      service.unsupported("Biological sex")
    }

    AsyncFunction("getBloodType") {
      service.unsupported("Blood type")
    }

    AsyncFunction("getDateOfBirth") {
      service.unsupported("Date of birth")
    }

    AsyncFunction("getFitzpatrickSkinType") {
      service.unsupported("Fitzpatrick skin type")
    }

    AsyncFunction("getWheelchairUse") {
      service.unsupported("Wheelchair use")
    }

    AsyncFunction("enableBackgroundDelivery") { _: String, _: Int ->
      service.unsupported("Background delivery")
    }

    AsyncFunction("disableBackgroundDelivery") { _: String ->
      service.unsupported("Background delivery")
    }

    AsyncFunction("disableAllBackgroundDelivery") {
      service.unsupported("Background delivery")
    }

    AsyncFunction("observeTypes") { _: List<String> ->
      service.unsupported("Observer queries")
    }

    AsyncFunction("clearObserverQueries") {
      Unit
    }
  }

  private suspend fun requestHealthPermissions(permissions: Set<String>): Set<String> {
    val activity: Activity = appContext.throwingActivity
    val intent = permissionContract.createIntent(activity, permissions)
    return withContext(Dispatchers.Main) {
      if (intent.action == RequestMultiplePermissions.ACTION_REQUEST_PERMISSIONS) {
        val platformPermissions =
          intent.getStringArrayExtra(RequestMultiplePermissions.EXTRA_PERMISSIONS)?.toSet()
            ?: permissions
        requestPlatformPermissions(platformPermissions)
      } else {
        requestHealthConnectActivityPermissions(activity, intent)
      }
    }
  }

  private suspend fun requestPlatformPermissions(permissions: Set<String>): Set<String> {
    val manager = appContext.permissions ?: throw HealthKitUnavailableException()
    return suspendCancellableCoroutine { continuation ->
      manager.askForPermissions(
        { responses ->
          if (continuation.isActive) {
            continuation.resume(
              responses.filter { it.value.status == PermissionsStatus.GRANTED }.keys.toSet()
            )
          }
        },
        *permissions.toTypedArray()
      )
    }
  }

  private suspend fun requestHealthConnectActivityPermissions(
    activity: Activity,
    intent: Intent
  ): Set<String> =
    suspendCancellableCoroutine { continuation ->
      permissionContinuation?.cancel()
      permissionContinuation = continuation
      continuation.invokeOnCancellation {
        if (permissionContinuation === continuation) {
          permissionContinuation = null
        }
      }
      try {
        activity.startActivityForResult(intent, REQUEST_PERMISSIONS)
      } catch (error: Exception) {
        permissionContinuation = null
        continuation.resumeWithException(error)
      }
    }

  companion object {
    private const val REQUEST_PERMISSIONS = 0x484B
  }
}

File: ios/ExpoHealthKitModule.swift
import ExpoModulesCore

public class ExpoHealthKitModule: Module {
  private let service = HealthKitService()

  public func definition() -> ModuleDefinition {
    Name("ExpoHealthKit")

    Events("onUpdate")

    OnCreate {
      self.service.onUpdate = { type in
        self.sendEvent("onUpdate", [
          "type": type
        ])
      }
    }

    OnDestroy {
      self.service.stopObserving()
    }

    Function("isHealthDataAvailable") {
      self.service.isAvailable()
    }

    AsyncFunction("requestAuthorization") { (options: AuthorizationOptions) in
      try await self.service.requestAuthorization(options)
    }

    AsyncFunction("getAuthorizationStatus") { (identifier: String) in
      try self.service.authorizationStatus(for: identifier)
    }

    AsyncFunction("getRequestStatusForAuthorization") { (options: AuthorizationOptions) in
      try await self.service.requestStatusForAuthorization(options)
    }

    AsyncFunction("queryQuantitySamples") { (options: QuantityQueryOptions) in
      try await self.service.queryQuantitySamples(options)
    }

    AsyncFunction("queryCategorySamples") { (options: CategoryQueryOptions) in
      try await self.service.queryCategorySamples(options)
    }

    AsyncFunction("queryWorkouts") { (options: WorkoutQueryOptions) in
      try await self.service.queryWorkouts(options)
    }

    AsyncFunction("queryElectrocardiograms") { (options: ElectrocardiogramQueryOptions) in
      try await self.service.queryElectrocardiograms(options)
    }

    AsyncFunction("queryActivitySummaries") { (options: ActivitySummaryQueryOptions) in
      try await self.service.queryActivitySummaries(options)
    }

    AsyncFunction("queryClinicalRecords") { (options: ClinicalRecordQueryOptions) in
      try await self.service.queryClinicalRecords(options)
    }

    AsyncFunction("queryAudiograms") { (options: DateRangeQueryOptions) in
      try await self.service.queryAudiograms(options)
    }

    AsyncFunction("queryWorkoutRoute") { (options: WorkoutRouteQueryOptions) in
      try await self.service.queryWorkoutRoute(options)
    }

    AsyncFunction("queryCorrelations") { (options: CorrelationQueryOptions) in
      try await self.service.queryCorrelations(options)
    }

    AsyncFunction("saveCorrelation") { (input: CorrelationInput) in
      try await self.service.saveCorrelation(input)
    }

    AsyncFunction("queryHeartbeatSeries") { (options: HeartbeatSeriesQueryOptions) in
      try await self.service.queryHeartbeatSeries(options)
    }

    AsyncFunction("queryStatistics") { (options: StatisticsQueryOptions) in
      try await self.service.queryStatistics(options)
    }

    AsyncFunction("queryStatisticsCollection") { (options: StatisticsCollectionQueryOptions) in
      try await self.service.queryStatisticsCollection(options)
    }

    AsyncFunction("queryAnchored") { (options: AnchoredQueryOptions) in
      try await self.service.queryAnchored(options)
    }

    AsyncFunction("saveQuantitySample") { (sample: QuantitySampleInput) in
      try await self.service.saveQuantitySample(sample)
    }

    AsyncFunction("saveCategorySample") { (sample: CategorySampleInput) in
      try await self.service.saveCategorySample(sample)
    }

    AsyncFunction("saveWorkout") { (workout: WorkoutInput) in
      try await self.service.saveWorkout(workout)
    }

    AsyncFunction("deleteObjects") { (options: DeleteObjectsOptions) in
      try await self.service.deleteObjects(options)
    }

    AsyncFunction("getBiologicalSex") {
      try self.service.biologicalSex()
    }

    AsyncFunction("getBloodType") {
      try self.service.bloodType()
    }

    AsyncFunction("getDateOfBirth") {
      try self.service.dateOfBirth()
    }

    AsyncFunction("getFitzpatrickSkinType") {
      try self.service.fitzpatrickSkinType()
    }

    AsyncFunction("getWheelchairUse") {
      try self.service.wheelchairUse()
    }

    AsyncFunction("enableBackgroundDelivery") { (type: String, frequency: Int) in
      try await self.service.enableBackgroundDelivery(type: type, frequency: frequency)
    }

    AsyncFunction("disableBackgroundDelivery") { (type: String) in
      try await self.service.disableBackgroundDelivery(type: type)
    }

    AsyncFunction("disableAllBackgroundDelivery") {
      try await self.service.disableAllBackgroundDelivery()
    }

    AsyncFunction("observeTypes") { (types: [String]) in
      try self.service.startObserving(types)
    }

    AsyncFunction("clearObserverQueries") {
      self.service.stopObserving()
    }

    OnStopObserving("onUpdate") {
      self.service.stopObserving()
    }
  }
}
